splainer-search.js

```diff
@@ -1037,6 +1038,62 @@ angular.module('o19s.splainer-search')
 
 'use strict';
 
+angular.module('o19s.splainer-search')
+  .service('queryTemplateSvc', function queryTemplateSvc() {
+    var self      = this;
+    self.hydrate = hydrate;
+
+    var defaultConfig = {
+      encodeURI: false,
+      defaultKw: '""',
+    };
+
+    function encode(queryPart, config) {
+      if (config.encodeURI) {
+        return encodeURIComponent(queryPart);
+      } else {
+        return queryPart;
+      }
+    }
+
+
+    function keywordMapping(queryText) {
+      var queryTerms    = queryText.split(/[ ,]+/);
+      var maxKeywords = 10;
+      var numTerms = queryTerms.length;
+      for (var i = numTerms; i < maxKeywords; i++) {
+        queryTerms.push(null);
+      }
+      return queryTerms;
+    }
+
+    function hydrate(template, queryText, config) {
+      if (!config) {
+        config = defaultConfig;
+      }
+
+      if (queryText === null || angular.isUndefined(queryText)) {
+        return template;
+      }
+
+      var replaced  = template.replace(/#\$query##/g, encode(queryText, config));
+      var idx = 0;
+      angular.forEach(keywordMapping(queryText), function(queryTerm) {
```

Comment:
It's a little more complex than that

1. I want to set an upper bound that's reasonable, someone could put
#$query10000## and screw up
2. For any keywords missing, say #$query5## for a query "dog cat" I
actually need to insert default values that will have no impact ("") in
Solr or ES

But I would probably do #$query\d## and just do one digit, which could be a
performance improvement. However then I'd have to figure out how to replace
each one.

-Doug

On Wed, Dec 2, 2015 at 9:50 AM, Youssef Chaker <notifications@github.com>
wrote:

> In splainer-search.js
> <https://github.com/o19s/splainer-search/pull/22#discussion_r46422410>:
>
> > +      }
> > +      return queryTerms;
> > +    }
> > +
> > +    function hydrate(template, queryText, config) {
> > +      if (!config) {
> > +        config = defaultConfig;
> > +      }
> > +
> > +      if (queryText === null || angular.isUndefined(queryText)) {
> > +        return template;
> > +      }
> > +
> > +      var replaced  = template.replace(/#\$query##/g, encode(queryText, config));
> > +      var idx = 0;
> > +      angular.forEach(keywordMapping(queryText), function(queryTerm) {
>
> can we flip the logic behind the loop?
> Determine how many /#\$query\d+##/ match, that gives you the number n of
> queries that need to be replaced.
> Loop n times, replace var regex = new RegExp('#\\$query' + (i) + '##',
> 'g'); with terms[i]
> This way you won't have to set var maxKeywords = 10; and you won't have
> to fill the array with nulls
> Less looping == faster code.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/o19s/splainer-search/pull/22/files#r46422410>.
>